DocumentManager

hawken93 opened the pull request:
Making DocumentManager and taking the functions that deal with DOM in there.

Also fixed:
* Image preloading
* less DOM work for ccast audio
* Size of poster image in details page
* linting

YouKnowBlom commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,533 @@
+// Duration between each backdrop switch
+const backdropPeriodMs: number = 30000;
+
+// DocumentManager should never ha
```
Ha what?

hawken93 replied:
Probably an unfinished justification for having it as a static class

simonmarklar commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,532 @@
+    private static getPrimaryImageUrl(item: BaseItemDto): string | null {
+        if (item.AlbumPrimaryImageTag)
+            return JellyfinApi.createUrl(
+                'Items/' +
```
If you didnt know, there's a relatively new feature in JS you can use to make this (imho) more legible than string concatenation - "Template Literals" - so you can ditch the string concat and the return statement would become

```
return JellyfinApi.createUrl(`Items/${item.AlbumId}/Images/Primary?tag=${item.AlbumPrimaryImageTag}`);
```

simonmarklar commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,532 @@
+    public static showItemId(itemId: string): void {
+        const url = JellyfinApi.createUserUrl('Items/' + itemId);
+
+        JellyfinApi.authAjax(url, 'GET').then(function (item: BaseItemDto) {
```
might not be a big deal in this case but its probably worthwhile returning this promise so whatever calls this function can wait for it to finish and be sure that the dom changes have been applied

simonmarklar commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,532 @@
+    private static getRatingHtml(item: BaseItemDto): string {
+        var html: string = '';
+        if (item.CommunityRating != null) {
+            html +=
```
highly recommend using template literals for html, its pretty much the main reason they exist :)

simonmarklar commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,532 @@
+// Duration between each backdrop switch in ms
+const backdropPeriodMs: number = 30000;
+
+export abstract class DocumentManager {
```
Sorry if I am missing something, but what value does making a class with all static private/public methods/props bring vs. a typescript module which exports the public methods (but as functions)? 

When I first started reading this I assumed because it was an abstract class there would be another class which implements it. Instead its pretty much doing the job of a module but with an extra cognitive load.

hawken93 replied:
I've tried to write a long and complicated answer, but to make it easy: It's functionally the same but I find it more intuitive when it's organized into an object. For me, it decreases cognitive load rather than increase it.

In this case, before the calls into DocumentManager, you can see in the other files just by looking at the calls that the functions are inside this class. By making sure all html manipulations are in this class, you'll see by the "DocumentManager" prefix that this call does a task related to html manipulation.
I know that you can kind of mess with the imports to get something of the same effect, but I just like this style better.

hawken93 replied:
Also I should probably move the adjacent const into the class just to avoid mixing OOP and non-OOP

hawken93 replied:
But mostly this is personal preference

simonmarklar replied:
yeah its functionally the same, I guess my main problem with it is that there is no concept of an abstract class in javascript. Even classes are really just sugar on top of `Function`. But that is my personal preference and there's no technical reason to choose one way over the other. I was just curious as to why you chose to use a class, Perhaps there was something I wasn't aware of that i could learn from

I do like that you've brought everything DOM related into one place, brings more structure to the project :)

hawken93 replied:
I was thinking about my decision to leave the backdropPeriodMs variable outside the class.

I would ideally make this something configurable. So that the user connecting to this chromecast can set this property to something that suits them. Or disable it completely.

We also have some user over at reddit who would prefer to switch the whole idle background image. In this case it would make sense to be able to pull this from the server.

I'll refactor this to a setter/getter style thing so it can be done easily later on :)

hawken93 commented:
Thanks for the review! I'll get back to it :)

hawken93 commented:
@simonmarklar I've done changes you suggested and I think I'm ready for another review :)

hawken93 commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,512 @@
+        }, this.backdropPeriodMs);
+    }
+
+    // null disables backdrops
```
TODO here is to clear the backdrop if it gets set to null. Relatively minor issue though.

simonmarklar commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,511 @@
+        var html: string = '';
+        if (item.CommunityRating != null) {
+            html +=
+                `<div class="starRating" title="${item.CommunityRating}"></div>` +
```
there's no need for string concatenation with template literals - they handle new lines just fine :)  eg:

```
`<div class="starRating" title="${item.CommunityRating}"></div>
<div class="starRatingValue">${item.CommunityRating.toFixed(
  1
)}</div>`
```

hawken93 replied:
I've done this previously, but I found the way that prettier split toFixed so obnoxious that I decided to do this to get my own line breaks.

hawken93 replied:
Fixed it by moving this stuff straight into html

hawken93 commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,611 @@
+            return Promise.resolve();
+        }
+
+        this.backdropTimer = <any>(
```
This makes the linter unhappy, but after searching about the setInterval function, I came to the conclusion that it's a mess and I would need help to fix this

hawken93 replied:
I think we can merge despite this. It's just the only linter warning I couldn't remove.

ferferga replied:
This is a number, TypeScript is not wrong about the return type: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setInterval#return_value

hawken93 replied:
Typescript thinks this is not a number

hawken93 replied:
```
ERROR in /home/hawken/git/mirror/jellyfin-chromecast/src/components/documentManager.ts
./components/documentManager.ts
[tsl] ERROR in /home/hawken/git/mirror/jellyfin-chromecast/src/components/documentManager.ts(352,9)
      TS2322: Type 'Timeout' is not assignable to type 'number'.
 @ ./components/maincontroller.ts 8:0-52 91:8-45
 @ ./app.ts 2:0-37
```

hawken93 replied:
https://stackoverflow.com/questions/45802988/typescript-use-correct-version-of-settimeout-node-vs-window

hawken93 replied:
This issue is beyond me, I really don't want to do some other hack instead of this hack. If we are going for a hack I'll add a comment here to tell the linter to leave it alone.

hawken93 commented:
Ready for reviews and regression testing!

Afaik this should behave no differently from master

ferferga commented:
@hawken93 Not strictly related to this PR but I think it fits to be added here:

Right now we "render" the UI for all devices, even Chromecast Audios. Can you add a check here, so if the device is a Chromecast Audio, a ``display: none`` is added to the whole page?

See https://developers.google.com/cast/docs/audio

hawken93 commented:
Hmm I could try. I'm thinking to mess with the detection to make it think my gen2 is an audio for test

hawken93 commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,611 @@
+            );
+        } else {
+            (<HTMLElement>element).style.backgroundImage =
+                src != null ? `url(${src})` : '';
```
I don't really like this because the new image loading is visible. I would prefer to preload the new src and then do something like a buffer swap. Someone better at frontend stuff that could help me with this?

ferferga replied:
@hawken93 I think this is what you want:

https://github.com/jellyfin/jellyfin-vue/pull/408/files#diff-4d53a82f6d48248fa1a22bd2f449b0604526d455c84d8b528ab723182b8f4c76R20

hawken93 replied:
Cool! I'll see what I can do

hawken93 replied:
Right. Ty :+1:

hawken93 commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,611 @@
+            }
+        }
+
+        let detailImageUrl = this.getPrimaryImageUrl(item);
+
+        if (
+            item?.UserData?.PlayedPercentage &&
+            item?.UserData?.PlayedPercentage < 100 &&
+            !item.IsFolder
+        ) {
+            this.setHasPlayedPercentage(false);
+            this.setPlayedPercentage(item.UserData.PlayedPercentage);
+
+            if (detailImageUrl != null)
+                detailImageUrl +=
+                    '&PercentPlayed=' +
+                    item.UserData.PlayedPercentage.toString();
+        } else {
+            this.setHasPlayedPercentage(false);
+            this.setPlayedPercentage(0);
+        }
+
+        this.setDetailImage(detailImageUrl);
```
This is the most central code that is responsible for the image on the left hand in the details page. Apparently this image is about half size of what it should be in some cases.

hawken93 replied:
Okay I found a css difference between stable and this, turns out there was a css messup.

hawken93 commented:
@ferferga I've eased up the DOM operations on cc audio devices :)

hawken93 commented:
I have updated the style a bit. I removed the extra h2 text in the waiting container, decreased the h1 to 45px and added a text container bottom right that gets updated with the name of the backdrop item. I've made it so that the text gets updated only when the image has loaded, so that the text switches in sync with the image.

hawken93 commented:
http://puu.sh/H3piL/254a5ec5ae.png

hawken93 commented:
![bilde](https://user-images.githubusercontent.com/1411903/103479702-ca262a80-4dcf-11eb-94fb-9b4d0bf62b13.png)

hawken93 commented:
![example](https://user-images.githubusercontent.com/1411903/103479711-dad6a080-4dcf-11eb-987f-d7a084b45d93.png)

hawken93 commented:
Requesting review and merge. I think this one has worked fine in mine and hamburglar's testing

hamburglar2160 commented:
I've been using this for the past week with no problems. This should get merged when somebody gets a moment to look at it

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+     * @returns {string | null} url to primary image
+     */
+    private static getPrimaryImageUrl(item: BaseItemDto): string | null {
+        if (item.AlbumPrimaryImageTag)
+            return JellyfinApi.createUrl(
+                `Items/${item.AlbumId}/Images/Primary?tag=${item.AlbumPrimaryImageTag}`
+            );
+        else if (item.ImageTags?.Primary)
+            return JellyfinApi.createUrl(
+                `Items/${item.Id}/Images/Primary?tag=${item.ImageTags.Primary}`
+            );
+        else return null;
```
We should probably add a rule for this, but imo all the blocks should have braces. **Don't fix it now, we can do it automatically with eslint in another PR**.

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+            detailRating.innerHTML = this.getRatingHtml(item);
+        }
+
+        const playedIndicator = document.getElementById('playedIndicator');
+
+        if (playedIndicator) {
+            if (item?.UserData?.Played) {
+                playedIndicator.style.display = 'block';
+                playedIndicator.innerHTML =
+                    '<span class="glyphicon glyphicon-ok"></span>';
+            } else if (item?.UserData?.UnplayedItemCount) {
+                playedIndicator.style.display = 'block';
+                playedIndicator.innerHTML = item.UserData.UnplayedItemCount.toString();
+            } else {
+                playedIndicator.style.display = 'none';
+            }
+        }
```
**This one requires changes**: First of all, I think we should add all the relevant HTML tags to the HTML file itself. The ones that won't appear by default (like this one) can have a ``d-none`` class. I would also create another ``d-block`` css class for ``display: block`` property.

Then, I would create a function like ``toggleVisibility`` which takes an HTMLElement as a parameter and toggles the ``d-none`` class. You're already setting ``display: none`` on initialize() and here, so you can reuse that function. You can also make one for the ``d-block``.

This would lead to better readability.

hawken93 replied:
we have a .hide class, not sure if it's used

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+        let html = '';
+        if (item.CommunityRating != null) {
+            html +=
+                `<div class="starRating" title="${item.CommunityRating}"></div>` +
```
Same as above

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+        if (item.CriticRating != null) {
+            const verdict = item.CriticRating >= 60 ? 'fresh' : 'rotten';
+            html +=
+                `<div class="${verdict} rottentomatoesicon" title="${verdict}"></div>` +
+                `<div class="criticRating">${item.CriticRating}%</div>`;
```
Same

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+        if (item.CommunityRating != null) {
+            html +=
+                `<div class="starRating" title="${item.CommunityRating}"></div>` +
+                '<div class="starRatingValue">' +
+                item.CommunityRating.toFixed(1) +
+                '</div>';
```
Same

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+                item.ParentBackdropImageTags.length
+            ) {
+                // otherwise get first backdrop from parent
+                src = JellyfinApi.createUrl(
+                    `Items/${item.ParentBackdropItemId}/Images/Backdrop/0?tag=${item.ParentBackdropImageTags[0]}`
+                );
```
Can't we also have a function that takes an ID, Image Type and tag and returns us this kind of url without needing it to be scattered all around?

hawken93 replied:
It seems like this one just takes backdrop number 0, ideally we would pick a random one, I'd like to express that too, somehow

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+        let backdropUrl: string | null = null;
+
+        if (item.BackdropImageTags && item.BackdropImageTags.length) {
+            backdropUrl = JellyfinApi.createUrl(
+                `Items/${item.Id}/Images/Backdrop/0?tag=${item.BackdropImageTags[0]}`
+            );
```
Same as above

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,678 @@
+            item.IndexNumber != null &&
+            item.ParentIndexNumber != null
+        ) {
+            let episode = `S${item.ParentIndexNumber}, E${item.IndexNumber}`;
```
Unrelated, but we should remember about this so we translate it once #111 is merged

ferferga commented on src/index.html:
```diff
@@ -18,9 +18,9 @@
             <img class="logo" src="img/logo.svg" />
-            <h1>Ready to cast</h1>
-            <h2>Select your media in Jellyfin and play it here</h2>
+            Ready to cast
```
Why remove ``h1``?

hawken93 replied:
It forces newline, alternately I could have tried the image within a h1 tag, but ultimately the css styling provides everthing we need

hawken93 commented:
Okay @hamburglar2160 request new test, I've only deployed this PR right now.

@ferferga  @simonmarklar new review?

hawken93 commented:
I've modified waiting and details pages to use relative scaling, should make 4k display look the same as 720p display and I think it's for the better, @hamburglar2160 opinion?

ferferga commented on src/components/deviceprofileBuilder.ts:
```diff
@@ -41,11 +41,11 @@ let profileOptions: ProfileOptions;
- * @param Property What property the condition should test.
- * @param Condition The condition to test the values for.
- * @param Value The value to compare against.
- * @param [IsRequired=false]
```
Why the default value is removed?

hawken93 replied:
I didn't understand how this would fit in to the jsdoc syntax, from the other functions I didn't see any practice of documenting default values with this syntax. Feel free to suggest a change

hawken93 replied:
Okay, fixed this :)

ferferga commented on src/components/documentManager.ts:
```diff
@@ -0,0 +1,778 @@
+     * on the details page. This happens when no media is playing,
+     * and the connected client is browsing the library.
+     *
+     * @param {BaseItemDto} item to show information about
+     * @returns {Promise<void>} for the page to load
```
JSDoc syntax is like this: @param {type} argumentInFunction - Description

@param {type} [argumentInFunction=value] - Description

for arguments that have default values.

Non-blocking though, I can fix that in the linting PR.

hawken93 replied:
As far as I can see the hyphen is optional and just a readability thing, but from this I can fix the other one

ferferga reviewed (approved):
Assumming this was fully tested after the new changes, this LGTM
